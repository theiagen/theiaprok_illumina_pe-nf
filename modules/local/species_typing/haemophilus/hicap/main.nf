process HICAP {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/biocontainers/hicap:1.0.3--py_0"

    input:
    tuple val(meta), path(assembly)
    val min_gene_percent_coverage
    val min_gene_percent_identity
    val min_broken_gene_percent_identity
    val broken_gene_length

    output:
    tuple val(meta), path("*.hicap.tsv")      , emit: hicap_results_tsv
    tuple val(meta), path("*_value.txt")      , emit: hicap_value_results
    path "versions.yml"                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def min_broken_gene_percent_identity_arg = min_broken_gene_percent_identity ?: 0.80
    def broken_gene_length_arg = broken_gene_length ?: 60
    def min_gene_percent_coverage_arg = min_gene_percent_coverage ?: 0.80
    def min_gene_percent_identity_arg = min_gene_percent_identity ?: 0.70
    
    """
    mkdir output_dir
    
    # Run HiCAP
    hicap \\
        -q ${assembly} \\
        -o output_dir \\
        --gene_coverage ${min_gene_percent_coverage_arg} \\
        --gene_identity ${min_gene_percent_identity_arg} \\
        --broken_gene_length ${broken_gene_length_arg} \\
        --broken_gene_identity ${min_broken_gene_percent_identity_arg} \\
        --threads ${task.cpus} \\
        ${args}
    
    filename=\$(basename ${assembly})
    
    # If there are no hits for a cap locus, no file is produced
    if [ ! -f ./output_dir/\${filename%.*}.tsv ]; then
        echo "No_hits" > hicap_serotype_value.txt
        echo "No_hits" > hicap_genes_value.txt
        touch ${prefix}.hicap.tsv
    else
        tail -n1 ./output_dir/\${filename%.*}.tsv | cut -f 2 > hicap_serotype_value.txt
        tail -n1 ./output_dir/\${filename%.*}.tsv | cut -f 4 > hicap_genes_value.txt
        mv ./output_dir/\${filename%.*}.tsv ${prefix}.hicap.tsv
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        hicap: \$(hicap --version 2>&1 | sed 's/^hicap //')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.hicap.tsv
    echo "a" > hicap_serotype_value.txt
    echo "bexA" > hicap_genes_value.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        hicap: \$(hicap --version 2>&1 | sed 's/^hicap //')
    END_VERSIONS
    """
}