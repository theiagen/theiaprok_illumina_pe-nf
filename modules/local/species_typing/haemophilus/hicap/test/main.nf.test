nextflow_process {
    name "Test Process HICAP"
    script "../main.nf"
    process "HICAP"

    test("HICAP - test") {
        setup {
            run("GUNZIP") {
                script "../../../../../nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = [
                        [id:'test_sample'],
                        file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/020/736/045/GCF_020736045.1_ASM2073604v1/GCF_020736045.1_ASM2073604v1_genomic.fna.gz', checkIfExists: true)
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = 0.80
                input[2] = 0.70
                input[3] = 0.80
                input[4] = 60
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.hicap_results_tsv, 
                    process.out.hicap_serotype, 
                    process.out.hicap_genes, 
                    process.out.versions
                ).match() }
            )
        }
    }

    test("HICAP  - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/020/736/045/GCF_020736045.1_ASM2073604v1/GCF_020736045.1_ASM2073604v1_genomic.fna.gz', checkIfExists: true)
                ])
                input[1] = 0.80
                input[2] = 0.70
                input[3] = 0.80
                input[4] = 60
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}