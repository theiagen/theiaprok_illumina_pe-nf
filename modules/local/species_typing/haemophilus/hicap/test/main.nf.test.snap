{
    "HICAP  - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.hicap.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "hicap_serotype:md5,60b725f10c9c85c70d97880dfe8191b3"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "hicap_genes:md5,d10eb57a297f3f78c9ae3dfa2d1e2ec1"
                    ]
                ],
                "3": [
                    "versions.yml:md5,8f155331d348220ce0f7d6467340608a"
                ],
                "hicap_genes": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "hicap_genes:md5,d10eb57a297f3f78c9ae3dfa2d1e2ec1"
                    ]
                ],
                "hicap_results_tsv": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.hicap.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "hicap_serotype": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "hicap_serotype:md5,60b725f10c9c85c70d97880dfe8191b3"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8f155331d348220ce0f7d6467340608a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-25T19:48:58.340810523"
    },
    "HICAP - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.hicap.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "hicap_serotype:md5,6381cdea650a02cb1c4b4094632ef966"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "hicap_genes:md5,6381cdea650a02cb1c4b4094632ef966"
                ]
            ],
            [
                "versions.yml:md5,8f155331d348220ce0f7d6467340608a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-25T19:48:48.319993312"
    }
}